core/source_map.rs

```diff
@@ -72,20 +77,70 @@ impl<G: SourceMapGetter> SourceMapper<G> {
       ext_source_maps: Default::default(),
       getter,
       stashed_file_name: Default::default(),
+      maybe_module_map: None,
     }
   }
 
+  pub fn set_module_map(&mut self, module_map: Rc<ModuleMap>) {
+    assert!(self.maybe_module_map.replace(module_map).is_none());
+  }
+
   pub fn has_user_sources(&self) -> bool {
     self.getter.is_some()
   }
 
+  pub fn apply_source_map_from_module_map(
+    &mut self,
+    scope: &mut v8::HandleScope,
+    file_name: &str,
+    line_number: u32,
+    column_number: u32,
+  ) -> Option<SourceMapApplication> {
+    let module_map = self.maybe_module_map.as_ref()?;
+    let id = module_map.get_id(file_name, RequestedModuleType::None)?;
+
+    let module_handle = module_map.get_handle(id).unwrap();
+    let module = v8::Local::new(scope, module_handle);
+    let unbound_module_script = module.get_unbound_module_script(scope);
+    let maybe_source_mapping_url =
+      unbound_module_script.get_source_mapping_url(scope);
+
+    // TODO(bartlomieju): This should be the last fallback and it's only useful
+    // for eval - probably for `Deno.core.evalContext()`.
+    let maybe_source_url = unbound_module_script
+      .get_source_url(scope)
+      .to_rust_string_lossy(scope);
+    eprintln!("maybe source url {}", maybe_source_url);
+
+    if !maybe_source_mapping_url.is_string() {
+      return None;
+    }
```

Comment:
I think we should change the API in `rusty_v8` to return `v8::String` and do the check for `is_string()` there. I believe it can only be `v8::Undefined` or `v8::String`, other types are unreachable here.

@devsnek WDYT?

Reply:
any particular reason? imo it is nice that rusty_v8 generally sticks to v8 semantics, it makes referencing documentation and bugs more helpful.